rust/noosphere-core/src/tracing.rs

```diff
@@ -1,27 +1,390 @@
+
+    /// Initialize tracing-based logging throughout the Noosphere body of code,
+    /// as well as dependencies that implement tracing-based logging.
+    ///
+    /// Invoking this function causes logs to be rendered until the termination
+    /// of the program. The default behavior is for logs to be rendered to
+    /// stdout. If this function is never called, logs will never be rendered.
+    /// Invoking this function more than once has no effect.
+    ///
+    /// The function accepts an optional [NoosphereLog] preset configuration
+    /// that controls log filter level and display format.
+    ///
+    /// Although this function accepts an optional preset configuration, it is
+    /// also sensitive to specific environment variables. The following
+    /// environment variables may be used to configure log behavior:
+    ///
+    ///  - **`RUST_LOG`**: A comma-separated list of directives as described in
+    ///    [`tracing-subscriber`][1] and [`env-logger`][2] documentation
+    ///  - **`NOOSPHERE_LOG`**: An optional preset value interpretted as a
+    ///    [NoosphereLog]
+    ///  - **`NOOSPHERE_LOG_LEVEL`**: A specific filter level that, if set, is
+    ///    assigned to all of the [NOOSPHERE_LOG_LEVEL_CRATES]
+    ///  - **`NOOSPHERE_LOG_FORMAT`**: A specific format that, if set, is
+    ///    interpretted as a [NoosphereLogFormat] and configured as the format
+    ///    for log output
+    ///
+    /// The configuration semantics are intended to be backwards compatible with
+    /// those defined by `tracing-subscriber` and `env-logger`. If only
+    /// `RUST_LOG` is set, then that configuration is applied as normal. The
+    /// [NoosphereLogFormat::Verbose] format is used in this case.
+    ///
+    /// If some preset is given in the function invocation, that preset is
+    /// decomposed to determine how to further modify the log filter level and
+    /// format. The directives will be added to the ones prescribed by
+    /// `RUST_LOG`. If a `NOOSPHERE_LOG` environment variable is specified, it
+    /// will take precedence over preset given as an argument.
+    ///
+    /// Finally, if a `NOOSPHERE_LOG_LEVEL` and/or `NOOSPHERE_LOG_FORMAT`
+    /// environment variable is configured, they will take final precedence to
+    /// determine the log filter level and format respectively.
+    ///
+    /// [1]: https://docs.rs/tracing-subscriber/0.3.17/tracing_subscriber/filter/struct.EnvFilter.html#directives
+    /// [2]: https://docs.rs/env_logger/0.10.0/env_logger/#enabling-logging
+    pub fn initialize_tracing(noosphere_log: Option<NoosphereLog>) {
+        INITIALIZE_TRACING.call_once(|| {
+            if let Err(error) = initialize_tracing_subscriber(noosphere_log) {
+                println!("Failed to initialize tracing: {}", error);
+            }
+        });
+    }
+
+    fn initialize_tracing_subscriber(noosphere_log: Option<NoosphereLog>) -> anyhow::Result<()> {
```

Comment:
consider: returning `()`, making this function's signature match its infallibility

Reply:
It's not infallible, and this let's us roll all the errors up into one at the call site.